File: shared/src/main/java/de/dreier/mytargets/shared/models/db/EndImage.java
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.models.db;

import android.support.annotation.Nullable;

import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.ForeignKey;
import com.raizlabs.android.dbflow.annotation.ForeignKeyAction;
import com.raizlabs.android.dbflow.annotation.ForeignKeyReference;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.structure.BaseModel;

import de.dreier.mytargets.shared.AppDatabase;

@Table(database = AppDatabase.class)
public class EndImage extends BaseModel implements Image {

    @Nullable
    @Column(name = "_id")
    @PrimaryKey(autoincrement = true)
    public Long _id = -1L;

    @Nullable
    @Column
    public String fileName = "";

    @Nullable
    @ForeignKey(tableClass = End.class, references = {
            @ForeignKeyReference(columnName = "end", columnType = Long.class, foreignKeyColumnName = "_id", referencedGetterName = "getId", referencedSetterName = "setId")},
            onDelete = ForeignKeyAction.CASCADE)
    public Long endId;

    public EndImage() {
    }

    public EndImage(String imageFile) {
        fileName = imageFile;
    }

    /**
     * @return The name of the image file, which is placed inside the files directory of the app
     */
    @Nullable
    public String getFileName() {
        return fileName;
    }

    /**
     * @param fileName The name of the image file, which is placed inside the files directory of the app
     */
    public void setFileName(String fileName) {
        this.fileName = fileName;
    }
}

File: shared/src/main/java/de/dreier/mytargets/shared/models/db/EndImage.kt
/*
 * Copyright (C) 2018 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.models.db

import android.arch.persistence.room.Entity
import android.arch.persistence.room.ForeignKey
import android.arch.persistence.room.ForeignKey.CASCADE
import android.arch.persistence.room.PrimaryKey
import android.os.Parcel
import android.os.Parcelable
import de.dreier.mytargets.shared.models.Image

@Entity(
    foreignKeys = [
        ForeignKey(
            entity = End::class,
            parentColumns = ["id"],
            childColumns = ["endId"],
            onDelete = CASCADE
        )
    ]
)
data class EndImage(
    @PrimaryKey(autoGenerate = true)
    var id: Long = 0,

    override var fileName: String = "",

    var endId: Long? = null
) : Image, Parcelable {
    constructor(imageFile: String) : this(fileName = imageFile)

    constructor(source: Parcel) : this(
        source.readLong(),
        source.readString(),
        source.readValue(Long::class.java.classLoader) as Long?
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeLong(id)
        writeString(fileName)
        writeValue(endId)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<EndImage> = object : Parcelable.Creator<EndImage> {
            override fun createFromParcel(source: Parcel): EndImage = EndImage(source)
            override fun newArray(size: Int): Array<EndImage?> = arrayOfNulls(size)
        }
    }
}
